egs/wsj/s5/steps/cleanup/combine_short_segments.sh

```diff
@@ -0,0 +1,165 @@
+  echo ""
+  echo "Usage: steps/cleanup/combine_short_segments.sh [options] <minimum-length> <data-in> <data-out>"
+  echo "e.g.:  steps/cleanup/combine_short_segments.sh 2.0 data/train data/train_min2sec"
+  echo ""
+  echo "Options:"
+  echo "    --frame-shift   <frame-shift|10>         # Frame shift in milliseconds.  Only relevant if"
+  echo "                                             # <data-in>/segments does not exist (in which case"
+  echo "                                             # the lengths are to be worked out from the features)."
+  exit 1;
+fi
+
+#This script works in the unit of frames
+frame_per_sec=$(( 1000/$frame_shift ))
+min_seg_len=$(echo $1*$frame_per_sec | bc)
+input_dir=$2
+output_dir=$3
+
+for f in spk2utt text utt2spk feats.scp; do
+  [ ! -f $input_dir/$f ] && echo "$0: no such file $input_dir/$f" && exit 1;
+done
+
+export LC_ALL=C
+
+# This is the main function, to look for segments with length shorter than the specified length
+# and concatenates them with other segments to make sure every combined segments are 
+# with enough length. Here, after the search of the under-length segment, it looks for
+# another segment where the combined length is the closest to the specified length.
+
+function check_and_combine_utt {
+
+echo $min_seg_len |  perl -e '
+  $min_seg_len = <STDIN>;
+  ($u2s_in, $s2u_in, $feat_in, $len_in, $text_in, 
+    $u2s_out, $feat_out, $text_out) = @ARGV;
+  open(UI, "<$u2s_in") || die "Error: fail to open $u2s_in\n";
+  open(SI, "<$s2u_in") || die "Error: fail to open $s2u_in\n";
+  open(FI, "<$feat_in") || die "Error: fail to open $feat_in\n";
+  open(LI, "<$len_in") || die "Error: fail to open $len_in\n";
+  open(TI, "<$text_in") || die "Error: fail to open $text_in\n";
+  open(UO, ">$u2s_out") || die "Error: fail to open $u2s_out\n";
+  open(FO, ">$feat_out") || die "Error: fail to open $feat_out\n";
+  open(TO, ">$text_out") || die "Error: fail to open $text_out\n";
+  while (<UI>) {
+    chomp;
+    @col = split;
+    @col == 2 || die "Error: bad line $_\n";
+    ($utt_id, $spk) = @col;
+    $utt2spk{$utt_id} = $spk;
+  }
+  while (<SI>) {
+    chomp;
+    @col = split;
+    $spks = join(" ", @col[1..@col-1]);
+    $spk2utt{$col[0]} = $spks;
+  }
+  while (<FI>) {
+    chomp;
+    @col = split;
+    @col == 2 || die "Error: bad line $_\n";
+    ($utt_id, $feat) = @col;
+    $uttlist{$utt_id} = $utt_id;
+    $utt2feat{$utt_id} = $feat;
+  }
+  while (<LI>) {
+    chomp;
+    @col = split;
+    @col == 2 || die "Error: bad line $_\n";
+    ($utt_id, $len) = @col;
+    $utt2len{$utt_id} = $len;
+    $utt2item{$utt_id} = 1;
+  }
+  while (<TI>) {
+    chomp;
+    @col = split;
+    $text = join(" ", @col[1..@col-1]);
+    $utt2text{$col[0]} = $text;
+  }
+
+  foreach $seg (sort keys %uttlist) {
+    if ($utt2len{$seg} < $min_seg_len) {
+      $localmin = inf;
+      $target = $seg;
+      @utts = split(" ",$spk2utt{$utt2spk{$seg}});
+      foreach $seg2 (@utts) {
```

Comment:
I don't think it is good to combine non-contiguous segments. This will affect things like ivector estimation. Ensure that the seg2 is next to seg. Keep combining segments until you reach minimum length.